libraries/AP_UAVCAN/AP_UAVCAN.cpp

```diff
@@ -705,6 +762,135 @@ void AP_UAVCAN::do_cyclic(void)
         led_out_sem_give();
     }
 
+    _tunnel_sem->take_blocking();
+    tunnel_send();
+    _tunnel_sem->give();
+}
+
+bool AP_UAVCAN::tunnel_sem_take()
+{
+    bool sem_ret = _tunnel_sem->take(10);
+    if (!sem_ret) {
+        debug_uavcan(1, "AP_UAVCAN Tunnel semaphore fail\n\r");
+        _tunnel_stats.sem_take_fail++;
+    }
+    return sem_ret;
+}
+
+void AP_UAVCAN::tunnel_sem_give()
+{
+    _tunnel_sem->give();
+}
+
+/*
+ *  Transmit all bytes that have been queued in the outbound tunnel object.
+ *  returns false if send fails, usually due to UAVCAN tx buffers being full.
+ */
+bool AP_UAVCAN::tunnel_flush()
+{
+    if (_tunnel.bdcst_msg.buffer.size() == 0) {
+        _tunnel.resend = false;
+        return true;
+    }
+
+    int32_t result = tunnel_broadcast_array[_uavcan_i]->broadcast(_tunnel.bdcst_msg);
+
+    if (result >= 0) {
+        _tunnel.last_send_ms = AP_HAL::millis();
+        // we're re-using the same msg, so lets make sure to clear it
+        _tunnel.bdcst_msg.buffer.clear();
+        return true;
+    }
+
+    // send failed. Let's quit now and try again on the next cycle
```

Comment:
how often does this send fail? It should be very rare

Reply:
can't remember.. probably zero. It's being logged so should be easy to test if you run plane code

Reply:
I checked. It has alwalys been zero so far